### ISSUE TITLE
More on overlay performance

### ISSUE THREAD

**Issue by ozak** [geopandas/geopandas#400]
This is just a follow up to #338, but wanted to make sure someone sees my posts. I was trying to use ``overlay`` and noticed it is impossibly slow. So I ended up coding some functions to take care of this. Using the example in #338 I tested and the new functions are much faster, so I am wondering if there is interest and I could create a pull that improves performance. Here's the function (for now it only implements ``intersection`` and ``difference``, but I could generalize it):

    def spatial_overlays(df1, df2, how='intersection'):
        '''Compute overlay intersection of two 
            GeoPandasDataFrames df1 and df2
        '''
        df1 = df1.copy()
        df2 = df2.copy()
        df1['geometry'] = df1.geometry.buffer(0)
        df2['geometry'] = df2.geometry.buffer(0)
        if how=='intersection':
            # Spatial Index to create intersections
            spatial_index = df2.sindex
            df1['bbox'] = df1.geometry.apply(lambda x: x.bounds)
            df1['histreg']=df1.bbox.apply(lambda x:list(spatial_index.intersection(x)))
            pairs = df1['histreg'].to_dict()
            nei = []
            for i,j in pairs.items():
                for k in j:
                    nei.append([i,k])
            
            pairs = gp.GeoDataFrame(nei, columns=['idx1','idx2'], crs=df1.crs)
            pairs = pairs.merge(df1, left_on='idx1', right_index=True)
            pairs = pairs.merge(df2, left_on='idx2', right_index=True, suffixes=['_1','_2'])
            pairs['Intersection'] = pairs.apply(lambda x: (x['geometry_1'].intersection(x['geometry_2'])).buffer(0), axis=1)
            pairs = gp.GeoDataFrame(pairs, columns=pairs.columns, crs=df1.crs)
            cols = pairs.columns.tolist()
            cols.remove('geometry_1')
            cols.remove('geometry_2')
            cols.remove('histreg')
            cols.remove('bbox')
            cols.remove('Intersection')
            dfinter = pairs[cols+['Intersection']].copy()
            dfinter.rename(columns={'Intersection':'geometry'}, inplace=True)
            dfinter = gp.GeoDataFrame(dfinter, columns=dfinter.columns, crs=pairs.crs)
            dfinter = dfinter.loc[dfinter.geometry.is_empty==False]
            return dfinter
        elif how=='difference':
            spatial_index = df2.sindex
            df1['bbox'] = df1.geometry.apply(lambda x: x.bounds)
            df1['histreg']=df1.bbox.apply(lambda x:list(spatial_index.intersection(x)))
            df1['new_g'] = df1.apply(lambda x: reduce(lambda x, y: x.difference(y).buffer(0), [x.geometry]+list(df2.iloc[x.histreg].geometry)) , axis=1)
            df1.geometry = df1.new_g
            df1 = df1.loc[df1.geometry.is_empty==False].copy()
            df1.drop(['bbox', 'histreg', new_g], axis=1, inplace=True)
            return df1

and the example 
 
    import geopandas as gpd
    world = gpd.read_file(gpd.datasets.get_path('naturalearth_lowres'))
    capitals = gpd.read_file(gpd.datasets.get_path('naturalearth_cities'))
    countries = world[['geometry', 'name']]
    countries = countries.to_crs('+init=epsg:3395')[countries.name!="Antarctica"]
    capitals = capitals.to_crs('+init=epsg:3395')
    capitals['geometry']= capitals.buffer(500000)

    %time gpd.overlay(countries, capitals, how='intersection')
    CPU times: user 36.5 s, sys: 357 ms, total: 36.8 s
    Wall time: 38.8 s

    %time spatial_overlays(countries, capitals, how='intersection')
    CPU times: user 1.53 s, sys: 11.3 ms, total: 1.54 s
    Wall time: 1.59 s

As you can see an major improvement in performance due to the use of the spatial index.

**Comment by nickeubank** [comment]
Hey @ozak thanks for this! That is a huge improvement. If it passes tests, we'd be delighted to integrate (I don't actually have pull authority, but I'm sure @jorisvandenbossche or @kjordahl would be happy to see an improvement here too). Want to write a PR?

**Comment by ozak** [comment]
@nickeubank I'd be happy to do so. Couple of questions:

1. Are there tests for ``overlay`` that I can use? Looking at the example that you posted and comparing the outcome with what I wrote I see it generates basically the same output, but the number of features is different. E.g., for New Zealand ``gpd.overlay`` has two features, while my function returns one feature (equivalent to a union of the two in ``gpd.overlay``).

2. Would it be preferable to add a new function ``spatial_overlay``? Looking at the current ``overlay`` function, which I do not fully understand yet, so replacing it completely may loose functionality...May also help for tests, in order to ensure the old ``overlay`` and the new one really generate the same results.

3. Not sure if the style etc. is similar to what is there already and I do not know  if there are code styling issues (saw something like that in one of the other issues discussions). E.g., I tend to use ``gp``, which I must have taken from older ``GeoPandas`` examples, but your example uses ``gpd``, which may be the newer style?

**Comment by nickeubank** [comment]
Hey @ozak : 

1) Yup, check out geopandas/geopandas/tests/test_overlay.py . There's a full suite in there for you to use. However, the one-feature / two-feature thing would definitely be a problem. Overlay is often used to get feature-specific attributes, so I definitely think it needs to return the right number of features. 

2) If you can pass all the tests referenced above, then it's fine to replace. That's the joy of tests! :)

3) I'd stick to the style you see in production code rather than examples. So for example, in `geopandas/geopandas/tools/overlay.py` we just direct-import the relevant tools and drop the library prefixes.

**Comment by jorisvandenbossche** [comment]
I answered in the related PR where @ozak first commented (https://github.com/geopandas/geopandas/pull/338), and to repeat it here again: I didn't merge those performance gains, because I first wanted to have better test coverage of `overlay`, to be sure my performance changes were actually still correct. 
But, I got stuck with the PR to add tests: https://github.com/geopandas/geopandas/pull/343 (there were some failures in the current code that I couldn't figure out then).

So basically, there is currently no decent coverage at all of the `overlay`  function. It only tests for the shape of the result (the number of rows), but does not test the actual result.

I think we first have to fix the test coverage, before improving the performance (how much I also want to do that, as I agree it is way to slow).

**Comment by nickeubank** [comment]
@jorisvandenbossche oops! sorry about that.

**Comment by jorisvandenbossche** [comment]
No problem! 
@ozak @nickeubank always welcome to take a look at https://github.com/geopandas/geopandas/pull/343. I should myself look back into it to remember again what the exact issue was. It uses the same examples from the docs as test cases (http://geopandas.readthedocs.io/en/latest/set_operations.html) and here is also a notebook with this: https://gist.github.com/jorisvandenbossche/d993c25e0ea8111778599990fb2da2bd

**Comment by nickeubank** [comment]
@jorisvandenbossche will take a look!

**Comment by ozak** [comment]
@nickeubank I will check the tests. 

Regarding point 1 above, I actually think the output in my implementation is what you should get and not what is being produced now (I imagine it is partly a matter of preferences since in principle one can go between them). In particular, I executed the same intersection in QGIS and it created the same output as in my implementation, which is what I would have expected (I do not have access to ArcGIS, but my impression is that it should create the same output also). See [here](https://www.dropbox.com/sh/3v2u450sbo5ws5a/AABHZ59-dgZAi6moNBQ_YYlIa?dl=0) (I have included the shape files and a script).

**Comment by nickeubank** [comment]
@ozak ah, I see your point on behavior. Yes, I believe you are correct that New Zealand should end up as a single feature. Sorry, I was thinking more of cases like Montevideo and Buenos Aires where we do want two features where those buffers overlap. You are correct re: New Zealand -- that was a single feature before and should stay one after. Thanks!

Working with @jorisvandenbossche on tests to improve those so you have more tests to work off. We should probably get that figured out first. feel free to weigh in -- #343

**Comment by nickeubank** [comment]
@ozak just occurred to me that the problems with the tests in #343 have more to do with bugs in current implementation than problems with tests. If you want to get going, feel free to just include the tests in that PR in your PR. jorvis does a lot, so might take him a while to tweak that.

**Comment by ozak** [comment]
Ok...I will try to find some time this week.

**Comment by nickeubank** [comment]
@ozak no rush! thanks!

**Comment by ozak** [comment]
I have created a commit that implements the improved function as discussed in #338 [a408ff1d98db81ae9b545d1d2d6b7f4451c597e3](ozak@a408ff1d98db81ae9b545d1d2d6b7f4451c597e3). I still have to see what the Travis test say, but they seem to be failing due to ``numpy`` related installation issues.

**Comment by ozak** [comment]
@nickeubank @jorisvandenbossche I am trying out one of the tests and it fails but I cannot understand why. Here's what I am running

```python
from __future__ import absolute_import

import tempfile
import shutil

from shapely.geometry import Point

from geopandas import GeoDataFrame, read_file
from geopandas.tests.util import unittest, download_nybb
from geopandas import overlay
N = 10
nybb_filename, nybb_zip_path = download_nybb()
nybb_filename
polydf = read_file(nybb_zip_path, vfs='zip://' + nybb_filename)
polydf
tempdir = tempfile.mkdtemp()
crs = {'init': 'epsg:4326'}
b = [int(x) for x in self.polydf.total_bounds]
b = [int(x) for x in polydf.total_bounds]
b
polydf2 = GeoDataFrame([
    {'geometry' : Point(x, y).buffer(10000), 'value1': x + y, 'value2': x - y}
    for x, y in zip(range(b[0], b[2], int((b[2]-b[0])/N)),
                    range(b[1], b[3], int((b[3]-b[1])/N)))], crs=crs)
polydf2
pointdf = GeoDataFrame([
    {'geometry' : Point(x, y), 'value1': x + y, 'value2': x - y}
    for x, y in zip(range(b[0], b[2], int((b[2]-b[0])/N)),
                    range(b[1], b[3], int((b[3]-b[1])/N)))], crs=crs)
polydf2.sindex
union_shape = (180, 7)
df = overlay(polydf, polydf2, how="union")
```

which gives the following error

```python
Exception ignored in: <bound method Handle.__del__ of <rtree.index.IndexStreamHandle object at 0x115e7b438>>
Traceback (most recent call last):
  File "/Users/bizcocho/Anaconda3/envs/GeoPandas/lib/python3.5/site-packages/rtree/index.py", line 875, in __del__
    self.destroy()
  File "/Users/bizcocho/Anaconda3/envs/GeoPandas/lib/python3.5/site-packages/rtree/index.py", line 863, in destroy
    if self._ptr is not None:
AttributeError: 'IndexStreamHandle' object has no attribute '_ptr'
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-34-9bfc272534df> in <module>()
----> 1 df = overlay(polydf, polydf2, how="union")

/Users/bizcocho/GitHub/geopandas/geopandas/tools/overlay.py in overlay(df1, df2, how, reproject)
    275         return dfsym
    276     elif how=='union':
--> 277         dfinter = overlay(df1, df2, how='intersection')
    278         dfsym = overlay(df1, df2, how='symmetric_difference')
    279         dfunion = dfinter.append(dfsym)

/Users/bizcocho/GitHub/geopandas/geopandas/tools/overlay.py in overlay(df1, df2, how, reproject)
    220         spatial_index = df2.sindex
    221         df1['bbox'] = df1.geometry.apply(lambda x: x.bounds)
--> 222         df1['sidx']=df1.bbox.apply(lambda x:list(spatial_index.intersection(x)))
    223         pairs = df1['sidx'].to_dict()
    224         nei = []

/Users/bizcocho/Anaconda3/envs/GeoPandas/lib/python3.5/site-packages/pandas/core/series.py in apply(self, func, convert_dtype, args, **kwds)
   2292             else:
   2293                 values = self.asobject
-> 2294                 mapped = lib.map_infer(values, f, convert=convert_dtype)
   2295 
   2296         if len(mapped) and isinstance(mapped[0], Series):

pandas/src/inference.pyx in pandas.lib.map_infer (pandas/lib.c:66124)()

/Users/bizcocho/GitHub/geopandas/geopandas/tools/overlay.py in <lambda>(x)
    220         spatial_index = df2.sindex
    221         df1['bbox'] = df1.geometry.apply(lambda x: x.bounds)
--> 222         df1['sidx']=df1.bbox.apply(lambda x:list(spatial_index.intersection(x)))
    223         pairs = df1['sidx'].to_dict()
    224         nei = []

AttributeError: 'NoneType' object has no attribute 'intersection'
```

But both the ``sindex`` exists and has the ``intersection`` function. The function works for ``difference`` and ``symmetric_difference`` if I first make both ``crs`` identical to ``EPSG:4326``

```python
polydf.to_crs(crs=crs, inplace=True)
```

It still fails when doing ``intersection``

```python
overlay(polydf, polydf2, how="intersection")
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
/Users/bizcocho/Anaconda3/envs/GeoPandas/lib/python3.5/site-packages/pandas/core/common.py in _asarray_tuplesafe(values, dtype)
    388                 result = np.empty(len(values), dtype=object)
--> 389                 result[:] = values
    390             except ValueError:

ValueError: could not broadcast input array from shape (0,12) into shape (0)

During handling of the above exception, another exception occurred:

ValueError                                Traceback (most recent call last)
/Users/bizcocho/Anaconda3/envs/GeoPandas/lib/python3.5/site-packages/pandas/core/frame.py in _ensure_valid_index(self, value)
   2463             try:
-> 2464                 value = Series(value)
   2465             except:

/Users/bizcocho/Anaconda3/envs/GeoPandas/lib/python3.5/site-packages/pandas/core/series.py in __init__(self, data, index, dtype, name, copy, fastpath)
    242                 data = _sanitize_array(data, index, dtype, copy,
--> 243                                        raise_cast_failure=True)
    244 

/Users/bizcocho/Anaconda3/envs/GeoPandas/lib/python3.5/site-packages/pandas/core/series.py in _sanitize_array(data, index, dtype, copy, raise_cast_failure)
   2951         else:
-> 2952             subarr = _asarray_tuplesafe(data, dtype=dtype)
   2953 

/Users/bizcocho/Anaconda3/envs/GeoPandas/lib/python3.5/site-packages/pandas/core/common.py in _asarray_tuplesafe(values, dtype)
    391                 # we have a list-of-list
--> 392                 result[:] = [tuple(x) for x in values]
    393 

ValueError: cannot copy sequence with size 12 to array axis with dimension 0

During handling of the above exception, another exception occurred:

ValueError                                Traceback (most recent call last)
<ipython-input-61-d95a422c850d> in <module>()
----> 1 overlay(polydf, polydf2, how="intersection")

/Users/bizcocho/GitHub/geopandas/geopandas/tools/overlay.py in overlay(df1, df2, how, reproject)
    229         pairs = pairs.merge(df1, left_on='idx1', right_index=True)
    230         pairs = pairs.merge(df2, left_on='idx2', right_index=True, suffixes=['_1','_2'])
--> 231         pairs['Intersection'] = pairs.apply(lambda x: (x['geometry_1'].intersection(x['geometry_2'])).buffer(0), axis=1)
    232         pairs = GeoDataFrame(pairs, columns=pairs.columns, crs=df1.crs)
    233         cols = pairs.columns.tolist()

/Users/bizcocho/Anaconda3/envs/GeoPandas/lib/python3.5/site-packages/pandas/core/frame.py in __setitem__(self, key, value)
   2417         else:
   2418             # set column
-> 2419             self._set_item(key, value)
   2420 
   2421     def _setitem_slice(self, key, value):

/Users/bizcocho/Anaconda3/envs/GeoPandas/lib/python3.5/site-packages/pandas/core/frame.py in _set_item(self, key, value)
   2482         """
   2483 
-> 2484         self._ensure_valid_index(value)
   2485         value = self._sanitize_column(key, value)
   2486         NDFrame._set_item(self, key, value)

/Users/bizcocho/Anaconda3/envs/GeoPandas/lib/python3.5/site-packages/pandas/core/frame.py in _ensure_valid_index(self, value)
   2464                 value = Series(value)
   2465             except:
-> 2466                 raise ValueError('Cannot set a frame with no defined index '
   2467                                  'and a value that cannot be converted to a '
   2468                                  'Series')

ValueError: Cannot set a frame with no defined index and a value that cannot be converted to a Series
```

I am not sure if all this is caused by the data. As shown in #338 using another dataset included in ``geopandas``, the function works without causing these errors. Any ideas?

**Comment by ozak** [comment]
Ok...just figured out what is happening. Definitely it is about the ``crs``. When they are both in the ``epsg:4326``, then there are no intersections, which the function actually does not know how to respond to. I will fix this.

**Comment by mullenkamp** [comment]
Thanks ozak. I tried out your function and it is substantially faster than the stock overlay function. I'm looking forward to having it included in the main repository.

**Comment by jliebrand** [comment]
Curious to get all these performance improvements... but it seems there hasn't been a lot of activity on this for some time. Is there still a plan to merge this in and get it released? Same for #338 ?

**Comment by Charlesfox1** [comment]
hi all, 
Thanks for posting the code for the speeded up intersection - seriously useful stuff! 
Two questions: 
1.) Has anyone seen analogous code for the 'union' operation of geopandas' overlay? I need this for a project I am working on.
2.) Is there still a plan to merge this?? It would be a big improvement over the current functionality, which isn't practical at scale.

**Comment by ozak** [comment]
@Charlesfox1 do you mean doing a union between two dataframes or union of elements of one only? The code also improved ``union`` of two dataframes as an operation, although the gains there were much smaller than in ``intersection``, if that is what you were referring to.

**Comment by Charlesfox1** [comment]
@ozak , thank you for getting back to me so quickly, on a thread that had not seen any activity for 6 months! 

In a nutshell, I have a single GeoDataFrame of polygon objects generated by buffering a roads shapefile. Many of these buffers overlap; I want to generate a GeoDataFrame (quickly!) that contains both the intersection and difference of every object - i.e. all 3 of the regions in the following diagram:

https://toblerity.org/shapely/code/intersection-sym-difference.hires.png

In your language above, I think that would be the union of all elements in a single DataFrame.

**Comment by Charlesfox1** [comment]
At the moment, I am doing this via: 
roads_buffer_union = gpd.overlay(roads, roads, how = 'union')

**Comment by ozak** [comment]
@Charlesfox1 that should be the right way to do it.

**Comment by Charlesfox1** [comment]
@ozak So to confirm, the useful code you wrote above cannot be used to speed up

roads_buffer_union = gpd.overlay(roads, roads, how = 'union') ?

If so....darn.

**Comment by Charlesfox1** [comment]
@ozak I note that the code above doesn't have a 'union' operation pre-programmed in it

**Comment by ozak** [comment]
Exactly as you wrote it, but you need the full new function (see #429).

**Comment by ozak** [comment]
Indeed, that was just the first trial. Check out pull request #429 or #338).

**Comment by awa5114** [comment]
Has this been integrated into the main repo? My intersections take forever to run

**Comment by jorisvandenbossche** [comment]
Yes, this was included in the 0.4.0 release. Would you be able to provide a reproducible example that illustrates the slow performance? (either if you can share the data, or otherwise try to generate some data that has similar characteristics)

**Comment by zedrdave** [comment]
Just adding my own observation that intersections are impossibly slow on the latest version: testing on a toy subset where boundary boxes have no overlap, it still takes a few seconds to check 5 pairs, which seems extremely abnormal…

**Comment by martinfleis** [comment]
@zedrdave Can you share your code? Do you have pygeos installed?

**Comment by zedrdave** [comment]
The code can be reduced to the single `overlay` call. I will try to provide a minimum working gdf (dealing with large OSM shapefiles, so I need to figure a way to extract a minimum series of geometries for which this happens).

I do have pygeos installed, and I am getting the warning about incompatible versions:
```
The Shapely GEOS version (3.8.0-CAPI-1.13.1 ) is incompatible with the GEOS version PyGEOS was compiled with (3.8.1-CAPI-1.13.3)
```

Could this be the cause?

Unfortunately, forcing an update of gpd to `dev` then gives me a lot more Shapely deprecation warnings…

**Comment by martinfleis** [comment]
@zedrdave That can be the issue. 

It is not about updating to dev version but fixing the environment to have both shapely and pygeos depend on the same version of GEOS. If you are using conda, installing both from `conda-forge` channel should resolve it.

Pygeos is still in an experimental phase now, so there are rough edges like this one. Those will be resolved in shapely2.0 (which includes pygeos).

**Comment by zedrdave** [comment]
@martinfleis Indeed, my bad: I just needed to reinstall both after making sure they were using the same version of GEOS.

Unfortunately, this did not seem to help with the speed problem.

I was able to build a reasonably-easy to reproduce 

```
# http://download.geofabrik.de/north-america/us/connecticut-latest-free.shp.zip    
ct_landuse = gpd.read_file(f'{path}/connecticut-latest-free/gis_osm_landuse_a_free_1.shp')

# http://download.geofabrik.de/north-america/us/arizona-latest-free.shp.zip
az_landuse = gpd.read_file(f'{path}/arizona-latest-free/gis_osm_landuse_a_free_1.shp')

# http://www2.census.gov/geo/tiger/TIGER2015/COUSUB/tl_2015_04_cousub.zip
arizona = gpd.read_file(f'{path}/tl_2015_04_cousub/tl_2015_04_cousub.shp')
az_county = arizona[arizona.COUNTYFP == '005'].dissolve('COUNTYFP')
az_county = az_county.to_crs(az_landuse.crs)

print(f"AZ county: {len(az_county)}\nCT land use: {len(ct_landuse)}\nAZ land use: {len(az_landuse)}")

# > AZ county: 1
# > CT land use: 20987
# > AZ land use: 125535

# no overlap between bounding boxes:

%timeit gpd.overlay(az_county, ct_landuse) 
# > 15.1 s ± 491 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

%timeit opt_overlay(az_county, ct_landuse)
# > 24.1 ms ± 233 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

As you can see, gpd takes a surprisingly long time, despite dealing with geometries that have no overlapping bounds.

Using a variant that removes non-overlapping bounds (see below for quick'n'dirty implementation) speeds things up greatly…

```
def pt_in_rect(x, y, r):
    return (r.minx <= x) & (x <= r.maxx) & (r.miny <= y) & (y <= r.maxy)

def bounds_overlap(df1, df2):
    b1 = df1.bounds.iloc[0]
    b2 = df2.bounds
    return (
            pt_in_rect(b2.minx, b2.miny, b1)
                | pt_in_rect(b2.minx, b2.maxy, b1)
                | pt_in_rect(b2.maxx, b2.miny, b1)
                | pt_in_rect(b2.maxx, b2.maxy, b1)
                | pt_in_rect(b1.minx, b1.miny, b2)
        )

def overlay(df1, df2):
    return gpd.overlay(df1, df2[bounds_overlap(df1, df2)])
```

**Comment by martinfleis** [comment]
What is `opt_overlay`? Do you have a code of that?

GeoPandas does not check for non-overlapping gdfs to return empty before passing the query to spatial index. That is expensive since you have 300k polygons. We do this check in `clip`, so it may be worth doing the same in `overlay`. That gives you back empty gdf in 160ms.

It takes time to build spatial index which you, in the end, don't use at all since bounding boxes are non-overlapping. So I don't think this is a performance issue with `overlay` as such, but with this specific case when we should do the prior check before building an index. I'll open an issue for that. Thanks for pointing it out!

**Comment by zedrdave** [comment]
The code for `opt_overlay` is at the bottom of my post. But it's just a very basic check for overlapping boxes.

This was just an example with a pathological case, but my real use-cases would include, say, 10% of overlapping (and 90% non-overlapping). In such cases, the optimisation might be worth it.

**Comment by martinfleis** [comment]
> In such cases, the optimisation might be worth it.

If you can come up with an idea how to further optimize `overlay` I am all ears. It has been optimised recently and I don't see much space for further optimisation at the moment (but I may be as well wrong).

**Comment by zedrdave** [comment]
I am currently playing with my data to see if I can improve it further (it's been a while since I've played with that sort of algorithms). My gut feeling is that it should be able to deal with clipping/overlaying 120k polygons much faster than it is (right now, it seems to take upward of an hour on my computer), but I may just need to accept it and look at other ways to speed up my particular date (`simplify` etc).

There might be another kinda low-hanging fruit with checking whether one polygon is fully enclosed in the other polygon's largest enclosed rectangle (log n)… 
This would majorly speed up cases like mine, where the clipping shape is fairly convex, and most of the regions being clipped are enclosed within that LR. But that's probably too specific to be useful here, and I suspect we'd quickly end up overlapping with the current algorithm.

**Comment by zedrdave** [comment]
@martinfleis Since we're here…
I've noticed a strange rounding issue with both `overlay` and `clip`:

```
coords = pd.Series([[(0,0),(0,1),(0.5,1),(0.5,0)], [(0.5,0),(0.5,1),(1,1),(1,0)]])

coords = coords.apply(lambda c: Polygon([(x-113,y+35) for x,y in c]))

gdf = gpd.GeoDataFrame(geometry=coords, crs=az_county.crs)

_, ((ax1, ax2), (ax3, ax4)) = plt.subplots(nrows=2, ncols=2, figsize=(8,8), sharex='all', sharey='all')
gdf.plot(ax=ax1)
gpd.clip(az_county, gdf.unary_union).plot(ax=ax2)
gpd.clip(gdf, az_county).plot(ax=ax3)
gpd.overlay(az_county, gdf).plot(ax=ax4)
plt.tight_layout()
```

![Unknown](https://user-images.githubusercontent.com/746979/99636485-c454cf80-2a43-11eb-875d-8815a1cca05b.png)

Is this kind of approximative behaviour to be expected, or should I log an issue for it?

**Comment by martinfleis** [comment]
@zedrdave Sorry, what is the issue here? I don't understand what do you mean by `rounding issue` and the result looks okay to me. Maybe open an issue if you think there is something going on.

**Comment by zedrdave** [comment]
@martinfleis sorry if that was unclear: both bottom plots should look like the upper-right one (ie, there's no reason the regions should have blank space separating them).
I'm unsure whether the problem is in the intersection algorithm, or the plotting…
I'm happy to open an issue… Just wanted to make sure this wasn't a trivial/expected behaviour.

**Comment by martinfleis** [comment]
@zedrdave There is no gap in reality, that is just how matplotlib plot polygons with zero linewidth. Geometrically, this is the correct output based on `bounds`.

**Comment by zedrdave** [comment]
@martinfleis but then, shouldn't the same gap show up when plotted in the upper-left plot?

**Comment by martinfleis** [comment]
@zedrdave That is caused by matplotlib's antialiasing. Not sure how exactly that works, but if you turn it off, there's no gap in any of them. As `gdf.plot(ax=ax1, antialiased=False)`.

**Comment by zedrdave** [comment]
@martinfleis Indeed! Completely missed out on that particular default value. Thanks for solving that riddle for me!

**Comment by zedrdave** [comment]
@martinfleis just a correction regarding my suggested change: (unsurprisingly) my quick-and-dirty 'overlap' function misses out on some cases (eg when the two boundary rectangles make a '+' shape). Using built-in `GeoSeries.intersects` should be fine.
